File: app/readplane/internal/snapshots/smart_mix.go
package snapshots

import (
	"errors"
	"fmt"
	"strings"
	"time"
)

const SmartMixProfileVersion = 1

var ErrSmartMixSnapshotStale = errors.New("Smart Mix snapshot requires FastAPI fallback")

// SmartMixProfileSummary is the compact, credential-free profile contract
// served by readplane v1. Full beat grids deliberately remain in FastAPI.
type SmartMixProfileSummary struct {
	TrackEntityUID       string    `json:"trackEntityUid"`
	ProfileVersion       int       `json:"profileVersion"`
	ProfileRevision      string    `json:"profileRevision"`
	Analyzer             string    `json:"analyzer"`
	AnalyzerVersion      string    `json:"analyzerVersion"`
	SourceRevision       string    `json:"sourceRevision"`
	DurationMS           int64     `json:"durationMs"`
	Quality              string    `json:"quality"`
	AnalyzedAt           time.Time `json:"analyzedAt"`
	BPM                  *float64  `json:"bpm,omitempty"`
	BPMConfidence        *float64  `json:"bpmConfidence,omitempty"`
	TempoStability       *float64  `json:"tempoStability,omitempty"`
	BeatAnchorMS         *int64    `json:"beatAnchorMs,omitempty"`
	DownbeatAnchorMS     *int64    `json:"downbeatAnchorMs,omitempty"`
	TimeSignature        *int64    `json:"timeSignature,omitempty"`
	BeatGridFormat       *string   `json:"beatGridFormat,omitempty"`
	Key                  *string   `json:"key,omitempty"`
	Scale                *string   `json:"scale,omitempty"`
	Camelot              *string   `json:"camelot,omitempty"`
	KeyConfidence        *float64  `json:"keyConfidence,omitempty"`
	IntroCueMS           *int64    `json:"introCueMs,omitempty"`
	OutroCueMS           *int64    `json:"outroCueMs,omitempty"`
	IntroLUFS            *float64  `json:"introLufs,omitempty"`
	OutroLUFS            *float64  `json:"outroLufs,omitempty"`
	TruePeakDBFS         *float64  `json:"truePeakDbfs,omitempty"`
	IntroEnergy          *float64  `json:"introEnergy,omitempty"`
	OutroEnergy          *float64  `json:"outroEnergy,omitempty"`
	IntroSpectralDensity *float64  `json:"introSpectralDensity,omitempty"`
	OutroSpectralDensity *float64  `json:"outroSpectralDensity,omitempty"`
	GlobalEnergy         *float64  `json:"globalEnergy,omitempty"`
	Danceability         *float64  `json:"danceability,omitempty"`
	Valence              *float64  `json:"valence,omitempty"`
	BlissVectorRevision  *string   `json:"blissVectorRevision,omitempty"`
}

// Validate rejects summaries that readplane v1 cannot safely represent.
// Callers fall back to FastAPI, which owns forward-compatible full profiles.
func (s SmartMixProfileSummary) Validate() error {
	if s.ProfileVersion != SmartMixProfileVersion {
		return fmt.Errorf(
			"%w: unsupported profile version %d",
			ErrSmartMixSnapshotStale,
			s.ProfileVersion,
		)
	}
	if strings.TrimSpace(s.TrackEntityUID) == "" ||
		strings.TrimSpace(s.ProfileRevision) == "" ||
		strings.TrimSpace(s.Analyzer) == "" ||
		strings.TrimSpace(s.AnalyzerVersion) == "" ||
		strings.TrimSpace(s.SourceRevision) == "" ||
		s.AnalyzedAt.IsZero() ||
		s.DurationMS < 0 {
		return fmt.Errorf(
			"%w: incomplete profile summary",
			ErrSmartMixSnapshotStale,
		)
	}
	switch s.Quality {
	case "full", "partial", "legacy", "unavailable":
	default:
		return fmt.Errorf(
			"%w: unsupported profile quality %q",
			ErrSmartMixSnapshotStale,
			s.Quality,
		)
	}
	return nil
}

File: app/crate/db/orm/smart_mix.py
from __future__ import annotations

from datetime import datetime

from sqlalchemy import (
    BIGINT,
    CheckConstraint,
    DateTime,
    Float,
    ForeignKey,
    Integer,
    LargeBinary,
    SmallInteger,
    Text,
    text,
)
from sqlalchemy.orm import Mapped, mapped_column

from crate.db.engine import Base


class TrackMixProfileRow(Base):
    __tablename__ = "track_mix_profiles"
    __table_args__ = (
        CheckConstraint("quality IN ('full', 'partial', 'legacy', 'unavailable')"),
        CheckConstraint("profile_version > 0"),
        CheckConstraint("bpm_confidence IS NULL OR bpm_confidence BETWEEN 0 AND 1"),
        CheckConstraint("tempo_stability IS NULL OR tempo_stability BETWEEN 0 AND 1"),
        CheckConstraint("key_confidence IS NULL OR key_confidence BETWEEN 0 AND 1"),
        CheckConstraint("(beat_grid_format IS NULL) = (beat_grid_data IS NULL)"),
    )

    track_id: Mapped[int] = mapped_column(
        BIGINT,
        ForeignKey("library_tracks.id", ondelete="CASCADE"),
        primary_key=True,
    )
    profile_version: Mapped[int] = mapped_column(Integer, nullable=False)
    profile_revision: Mapped[str] = mapped_column(Text, nullable=False)
    analyzer: Mapped[str] = mapped_column(Text, nullable=False)
    analyzer_version: Mapped[str] = mapped_column(Text, nullable=False)
    source_revision: Mapped[str] = mapped_column(Text, nullable=False)
    quality: Mapped[str] = mapped_column(Text, nullable=False)
    bpm: Mapped[float | None] = mapped_column(Float)
    bpm_confidence: Mapped[float | None] = mapped_column(Float)
    tempo_stability: Mapped[float | None] = mapped_column(Float)
    beat_anchor_ms: Mapped[int | None] = mapped_column(BIGINT)
    downbeat_anchor_ms: Mapped[int | None] = mapped_column(BIGINT)
    time_signature: Mapped[int | None] = mapped_column(SmallInteger)
    beat_grid_format: Mapped[str | None] = mapped_column(Text)
    beat_grid_data: Mapped[bytes | None] = mapped_column(LargeBinary)
    audio_key: Mapped[str | None] = mapped_column(Text)
    audio_scale: Mapped[str | None] = mapped_column(Text)
    key_camelot: Mapped[str | None] = mapped_column(Text)
    key_confidence: Mapped[float | None] = mapped_column(Float)
    intro_cue_ms: Mapped[int | None] = mapped_column(BIGINT)
    outro_cue_ms: Mapped[int | None] = mapped_column(BIGINT)
    intro_lufs: Mapped[float | None] = mapped_column(Float)
    outro_lufs: Mapped[float | None] = mapped_column(Float)
    true_peak_dbfs: Mapped[float | None] = mapped_column(Float)
    intro_energy: Mapped[float | None] = mapped_column(Float)
    outro_energy: Mapped[float | None] = mapped_column(Float)
    intro_spectral_density: Mapped[float | None] = mapped_column(Float)
    outro_spectral_density: Mapped[float | None] = mapped_column(Float)
    global_energy: Mapped[float | None] = mapped_column(Float)
    danceability: Mapped[float | None] = mapped_column(Float)
    valence: Mapped[float | None] = mapped_column(Float)
    bliss_vector_revision: Mapped[str | None] = mapped_column(Text)
    analyzed_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=text("NOW()"),
    )
